blender: implement multiple versions, move to fetchFromGitea

the-furry-hubofeverything opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Resolves #256845

I've changed the `fetchurl` to `fetchFromGitea`, to find the tagged sources easier (at the cost of download speed, I've found projects.blender.org isn't as fast as downloads.blender.org).

I ran into issues with applying the `draco.patch` to the addon repo, so I split it in two, patching each repo seperately.

I've roughly referenced `pkgs/minecraft-servers/default.nix` for my json to packages implementation, since I don't have much experience with nixpkgs. If that's not okay, please let me know.

**I moved a lot of the original code, but I kept it the same for the most part. I'm new to writing for nixpkgs, so I'm only implementing the parts I know and a major refactor of the original code is not within scope.**

- [x] Moved to fetchFromGitea
- [x] Implement version selection
- [x] Add `isLTS` attribute
- [x] Make `update.py`
- [x] Figure out `blender-with-packages`
- [x] Fix passthru test
- [x] Format code according to `CONTRIBUTING.md`
- [x] Added myself as maintainer
- [x] fix up packages dependent on `blender-with-packages`


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

the-furry-hubofeverything committed:
blender: Switch to fetchFromGitea

the-furry-hubofeverything committed:
blender: Added blender-addons repo

Fixes patching issue. Blender-addons-contrib is not included, as decided upstream.

the-furry-hubofeverything commented on pkgs/applications/misc/blender/versions.json:
```diff
@@ -0,0 +1,28 @@
+{
+
+    "testing": {
```
Forgot to remove, will do so later. Planning on adding stable versions only in this json.

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30874,15 +30874,15 @@ with pkgs;
+  blender-hip = blenderPackages-hip.latest;
 
-  blender-with-packages = callPackage ../applications/misc/blender/wrapper.nix { };
+  blenderPackages = callPackage ../applications/misc/blender { };
```
blenderPackages is not the attribute naming scheme for versioning. Look at how it's done in e.g. `boost`.

the-furry-hubofeverything replied:
Moved to `blender3_x`, `blender3_x-hip`, and with `blender`, `blender-lts`, `blender-hip` etc

veprbl replied:
Please see
https://nixos.org/manual/nixpkgs/stable/#sec-versioning for file organization
and last bullet of https://nixos.org/manual/nixpkgs/stable/#sec-package-naming for blender_3

the-furry-hubofeverything replied:
So `blender_3_x`, `blender-hip_3_x`?

veprbl replied:
The `_x` is omitted.

the-furry-hubofeverything replied:
> The `_x` is omitted.

I don't understand what you mean.

The example in the link you gave me (https://nixos.org/manual/nixpkgs/stable/#sec-package-naming) with their example has `json-c_0_9` and `json-c_0_11`.

If you mean `blender_3x`, personally I think that's confusing, given previously blender has versions like `2.93`, and blender uses these to indicate API breaking changes (major and minor versions).

the-furry-hubofeverything committed:
blender: implement multi version packages

moved derivation to generic.nix, passthrough `pythonPackages`

roughly referenced pkgs/minecraft-servers/default.nix

blender_3_3: init 3.3.14
blender_3_6: init 3.6.7
blender_4_0: init 4.0.2

the-furry-hubofeverything committed:
blender: implement multi version packages

moved derivation to generic.nix, passthrough `pythonPackages`

roughly referenced pkgs/minecraft-servers/default.nix

blender_3_3: init 3.3.14
blender_3_6: init 3.6.7
blender_4_0: init 4.0.2

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30874,15 +30874,23 @@ with pkgs;
+  ;
+
+  inherit (callPackages ../applications/misc/blender { hipSupport = true; })
+    blender3_3-hip
```
If I'm not mistaken, this pattern breaks `blender.override`.

the-furry-hubofeverything replied:
Changed to `inherit (recurseIntoAttrs`, would that do it? I don't know how to test for it.

veprbl replied:
To test, just try `nix-instantiate --expr 'with import <nixpkgs> {}; blender.override { inherit stdenv; }'`

the-furry-hubofeverything replied:
> No. There would need to be separate callPackages for each attribute.

Okay I've tested it, overrides should work. `callPackages` are created with each attribute in `default.nix`

the-furry-hubofeverything commented:
Almost ready. I just need to clean up the nix code.

the-furry-hubofeverything commented:
> Almost ready. I just need to clean up the nix code.

Completely forgot `wrapper.nix`. 

Okay, so, currently we have a couple of decisions to make.

## Version naming conventions

So far I've gotten three options

### all on top-level
Based on naming convention of `boost`, but using `inherit (recurseIntoAttrs` for specific versions. It has the advantage of not break overrides but it is expensive. 
Also, `all-packages.nix` needs to update every minor version change.

- blender
- blender-lts
- blender-hip
- blender-hip-lts
- blender_3_x
- blender-hip_3_x

### with `blenderVersions`(currently implmented)
Based on `nixVersions`, might remove `blender-hip-lts` and leave `blender-hip` only for backwards compatibility. Doesn't need changes to `all-packages.nix` for minor version updates. My preferred choice.

- blender
- blender-lts
- blender-hip
- blender-hip-lts
- blenderVersions
  - blender_3_x
  - blender-hip_3_x
  - lts
  - stable

### condensed `blenderVersions`
Users can define `blenderVersions.blender_3_x.override { hipSupport = true; };` if they want it. Generates the least amount of package entries.

- blender
- blender-lts
- blender-hip
- blender-hip-lts
- blenderVersions
  - blender_3_x
  - lts
  - stable


## blender-with-packages

From what I understand, `blender-with-packages` is a wrapper for Blender, but I don't understand how it works. 

I assume it allows other packages to work with Blender, like using `bpy`. I think, at it's current state, it only works with the `blender` alias. Would it be feasible to generate a wrapper for each Blender version (`blenderVersions.blender_3_x-with-packages`)?

But then for existing packages in nixpkgs (i'm only aware of `bpycv` so far), how would they implement their package for multiple versions of Blender?

Opinions? @lucasew (for `bpycv`) @veprbl @pennae @cillianderoiste

lucasew commented:
blender-with-packages basically do what python.with-packages does. Blender 2.x still uses python 3.9 and 3.x python 3.10 and we had a problem with linking which python packages are supported for each blender version. 

As far as I remember it takes blender as the first parameter and from a passthru attribute gets which python package set it supports (like python.pkgs).

lucasew commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,268 @@
+}:
+
+let
+  python = python310Packages.python;
```
I think its better to parametrize this pythonPackages thing and then expose it in passthrough.

All blender 3.x so far are based on python 3.10 but it may be not true in the near future then it would be a tech debt for the next touching this and a wrapper generator might get things wrong.

the-furry-hubofeverything replied:
Good idea.

the-furry-hubofeverything replied:
.... I've realized that I have no idea how to test `withPackages`, so I need your help. I have no idea if it works.

I also don't know how to help you modify your package so it works with the new system.

lucasew replied:
The same way you would test `python3.withPackages`, the only difference is that the blender derivation hints about which python version it's based and pass the package scope to the function you pass to withPackages. That's the reason it exists.

And to open a python REPL inside blender you run it with `blender --background --python-console`. You can also pass a script with `blender --python script.py` to check for the imports.

the-furry-hubofeverything commented:
`blender_3_3` doesn't pass render test, 

@amarshall you wrote the unit test, I don't know much about command-line use in blender, do you think there's an argument change or something I've missed?

the-furry-hubofeverything commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30874,15 +30874,23 @@ with pkgs;
-    stdenv = if stdenv.isDarwin then llvmPackages_10.stdenv else stdenv;
-    inherit (darwin.apple_sdk.frameworks) Cocoa CoreGraphics ForceFeedback OpenAL OpenGL;
-  };
+  inherit (recurseIntoAttrs (callPackage ../applications/misc/blender { }))
```
Fixed with `blenderVersions`, I hope.

the-furry-hubofeverything commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,268 @@
+  '';
+
+  passthru = {
+    inherit python isLTS pythonPackages withPackages;
```
@lucasew like this with `pythonPackages`? it passes through the specific `pythonPackages` that it uses, in this case an alias to `python310Packages`

lucasew replied:
Yes.

lucasew commented on pkgs/applications/misc/blender/wrapper.nix:
```diff
@@ -2,7 +2,7 @@
 , blender
 , makeWrapper
-, python3Packages
+, pythonPackages
```
Define the default value of `pythonPackages` as `blender.pythonPackages`, that way you can omit the pythonPackages parameter where you are calling this function.

lucasew replied:
`pythonPackages ? blender.pythonPackages`

Just check if it gets the default values right.

lucasew commented:
> `blender_3_3` doesn't pass render test,
> 
> @amarshall you wrote the unit test, I don't know much about command-line use in blender, do you think there's an argument change or something I've missed?

Which error?

the-furry-hubofeverything commented:
> > `blender_3_3` doesn't pass render test,
> > @amarshall you wrote the unit test, I don't know much about command-line use in blender, do you think there's an argument change or something I've missed?
> 
> Which error?

I forgot to include the log, sorry

```
copying path '/nix/store/5g9nilk7v6lv4gipybskiv91q36k8ww9-libelf-0.8.13' from 'https://cache.nixos.org'...
copying path '/nix/store/nm7al9rc9a81q5y7p942xh96n2smbg1j-llvm-15.0.7-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/qznjbcwwn39dnbd6rppr5gzsl2w916s3-libxshmfence-1.3.2' from 'https://cache.nixos.org'...
copying path '/nix/store/ja4ax8704yaap28a9v0xqbpx2ag8sckc-mesa-23.1.7' from 'https://cache.nixos.org'...
copying path '/nix/store/2j7b1ngdvqd0bidb6bn9icskwm6sq63v-perl-5.38.0' from 'https://cache.nixos.org'...
copying path '/nix/store/lf9pzvvg1y6kl4w89fh9cs73in0l5d39-stdenv-linux' from 'https://cache.nixos.org'...
copying path '/nix/store/17zxjnpgh8327zyfvbznn7p9r3nnykal-lm-sensors-3.6.0' from 'https://cache.nixos.org'...
copying path '/nix/store/swmp8c9a8indr65dkdmxjc1f26lgib40-mesa-23.1.7-drivers' from 'https://cache.nixos.org'...
Rendering with BLENDER_EEVEE...
Blender 3.3.11
could not get a list of mounted file-systems
GHOST/Wayland: error: XDG_RUNTIME_DIR is invalid or not set in the environment.
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x33ef963]
/nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(+0xba1a) [0x7ffff7f21a1a]
/nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(wl_display_connect+0x2e7) [0x7ffff7f1dc07]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c5b35a]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c49aad]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c485d9]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb298f1]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb12e97]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xc30eb5]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1963e09]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x19647c5]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1968797]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196ad50]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196bb69]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x586292]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x3307dde]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x534434]
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(+0x23ace) [0x7fffe9e3dace]
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(__libc_start_main+0x89) [0x7fffe9e3db89]
/nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x583ae5]
Unable to open a display
/build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 19:     7 Aborted                 (core dumped) /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender --background -noaudio --factory-startup --python-exit-code 1 --python scene-config.py --engine "$engine" --render-output "$out/$engine" --render-frame 1
```

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

lucasew commented:
> > > `blender_3_3` doesn't pass render test,
> > > @amarshall you wrote the unit test, I don't know much about command-line use in blender, do you think there's an argument change or something I've missed?
> > 
> > 
> > Which error?
> 
> I forgot to include the log, sorry
> 
> ```
> copying path '/nix/store/5g9nilk7v6lv4gipybskiv91q36k8ww9-libelf-0.8.13' from 'https://cache.nixos.org'...
> copying path '/nix/store/nm7al9rc9a81q5y7p942xh96n2smbg1j-llvm-15.0.7-lib' from 'https://cache.nixos.org'...
> copying path '/nix/store/qznjbcwwn39dnbd6rppr5gzsl2w916s3-libxshmfence-1.3.2' from 'https://cache.nixos.org'...
> copying path '/nix/store/ja4ax8704yaap28a9v0xqbpx2ag8sckc-mesa-23.1.7' from 'https://cache.nixos.org'...
> copying path '/nix/store/2j7b1ngdvqd0bidb6bn9icskwm6sq63v-perl-5.38.0' from 'https://cache.nixos.org'...
> copying path '/nix/store/lf9pzvvg1y6kl4w89fh9cs73in0l5d39-stdenv-linux' from 'https://cache.nixos.org'...
> copying path '/nix/store/17zxjnpgh8327zyfvbznn7p9r3nnykal-lm-sensors-3.6.0' from 'https://cache.nixos.org'...
> copying path '/nix/store/swmp8c9a8indr65dkdmxjc1f26lgib40-mesa-23.1.7-drivers' from 'https://cache.nixos.org'...
> Rendering with BLENDER_EEVEE...
> Blender 3.3.11
> could not get a list of mounted file-systems
> GHOST/Wayland: error: XDG_RUNTIME_DIR is invalid or not set in the environment.
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x33ef963]
> /nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(+0xba1a) [0x7ffff7f21a1a]
> /nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(wl_display_connect+0x2e7) [0x7ffff7f1dc07]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c5b35a]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c49aad]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c485d9]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb298f1]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb12e97]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xc30eb5]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1963e09]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x19647c5]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1968797]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196ad50]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196bb69]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x586292]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x3307dde]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x534434]
> /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(+0x23ace) [0x7fffe9e3dace]
> /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(__libc_start_main+0x89) [0x7fffe9e3db89]
> /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x583ae5]
> Unable to open a display
> /build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 19:     7 Aborted                 (core dumped) /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender --background -noaudio --factory-startup --python-exit-code 1 --python scene-config.py --engine "$engine" --render-output "$out/$engine" --render-frame 1
> ```

Using the `--background` should not give errors about unable to open display. It should run headless. It looks like a upstream bug. You can use xvfb-run to workaround this.

the-furry-hubofeverything commented:
Under linux, 

> > > > `blender_3_3` doesn't pass render test,
> > > > @amarshall you wrote the unit test, I don't know much about command-line use in blender, do you think there's an argument change or something I've missed?
> > > 
> > > 
> > > Which error?
> > 
> > 
> > I forgot to include the log, sorry
> > ```
> > copying path '/nix/store/5g9nilk7v6lv4gipybskiv91q36k8ww9-libelf-0.8.13' from 'https://cache.nixos.org'...
> > copying path '/nix/store/nm7al9rc9a81q5y7p942xh96n2smbg1j-llvm-15.0.7-lib' from 'https://cache.nixos.org'...
> > copying path '/nix/store/qznjbcwwn39dnbd6rppr5gzsl2w916s3-libxshmfence-1.3.2' from 'https://cache.nixos.org'...
> > copying path '/nix/store/ja4ax8704yaap28a9v0xqbpx2ag8sckc-mesa-23.1.7' from 'https://cache.nixos.org'...
> > copying path '/nix/store/2j7b1ngdvqd0bidb6bn9icskwm6sq63v-perl-5.38.0' from 'https://cache.nixos.org'...
> > copying path '/nix/store/lf9pzvvg1y6kl4w89fh9cs73in0l5d39-stdenv-linux' from 'https://cache.nixos.org'...
> > copying path '/nix/store/17zxjnpgh8327zyfvbznn7p9r3nnykal-lm-sensors-3.6.0' from 'https://cache.nixos.org'...
> > copying path '/nix/store/swmp8c9a8indr65dkdmxjc1f26lgib40-mesa-23.1.7-drivers' from 'https://cache.nixos.org'...
> > Rendering with BLENDER_EEVEE...
> > Blender 3.3.11
> > could not get a list of mounted file-systems
> > GHOST/Wayland: error: XDG_RUNTIME_DIR is invalid or not set in the environment.
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x33ef963]
> > /nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(+0xba1a) [0x7ffff7f21a1a]
> > /nix/store/9a9hfbv339z3jla9j59ql9q17wrs8sk1-wayland-1.22.0/lib/libwayland-client.so.0(wl_display_connect+0x2e7) [0x7ffff7f1dc07]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c5b35a]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c49aad]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1c485d9]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb298f1]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xb12e97]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0xc30eb5]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1963e09]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x19647c5]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x1968797]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196ad50]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x196bb69]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x586292]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x3307dde]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x534434]
> > /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(+0x23ace) [0x7fffe9e3dace]
> > /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6(__libc_start_main+0x89) [0x7fffe9e3db89]
> > /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender() [0x583ae5]
> > Unable to open a display
> > /build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 19:     7 Aborted                 (core dumped) /nix/store/vmf7cqhrgk6fj2qdpjdx05szxkhpfx4a-blender-3.3.11/bin/blender --background -noaudio --factory-startup --python-exit-code 1 --python scene-config.py --engine "$engine" --render-output "$out/$engine" --render-frame 1
> > ```
> 
> Using the `--background` should not give errors about unable to open display. It should run headless. It looks like a upstream bug. You can use xvfb-run to workaround this.

Under linux, headless EEVEE support only began from 3.4.
 [commit](https://projects.blender.org/blender/blender/commit/3195a381200eb98e6add8b0504f701318186946d)
 [changelog](https://wiki.blender.org/wiki/Reference/Release_Notes/3.4/EEVEE)
 
 I could probably write an exception for 3.3. I wonder if it would be too much trouble for them to backport it....

the-furry-hubofeverything commented:
I'm going to try `xvfb-run` first, then disabling it for version 3.3 and below if that doesn't work.

the-furry-hubofeverything commented:
1. Should I add myself as maintainer?
2. Should we add this to the 23.11 release notes? Especially for `blender.withPackages` and `blender.pythonPackages`

the-furry-hubofeverything committed:
blender: add hubble as maintainer

the-furry-hubofeverything commented:
It might need to be a separate PR, but does openpgl not build on aarch64?

the-furry-hubofeverything commented:
@cillianderoiste @veprbl ready for review I think

the-furry-hubofeverything commented:
> It might need to be a separate PR, but does openpgl not build on aarch64?

Fixed https://github.com/NixOS/nixpkgs/pull/260926

Though there might be additional dependencies broken with my trial cross-compilation

the-furry-hubofeverything commented:
@ofborg eval

the-furry-hubofeverything commented:
If #261521 is merged first, I'll rebase this PR. If this is merged first, then I'll close #261521.

the-furry-hubofeverything commented:
Did some quick formatting

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

the-furry-hubofeverything commented:
Updated blender_3_6 to 3.6.5, blender_3_3 to 3.3.12

the-furry-hubofeverything commented:
> ![Screenshot_20231020-104742](https://user-images.githubusercontent.com/15693688/276950460-7409cf58-5681-4641-a89f-f0f00c48ae1a.png)

see previous comment, working on it

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>blender (blender-lts ,blenderVersions.blender_3_6 ,blenderVersions.lts ,blenderVersions.stable)</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

the-furry-hubofeverything commented:
The update script is behaving really weirdly on my setup, probably because it's kind of not conventional (flake nixpath, workarounds). The script hangs on downloading things, and it would "download" over a few hours. 

Script may need further testing

the-furry-hubofeverything commented:
@lucasew hashes should be corrected now

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

the-furry-hubofeverything commented:
> @cillianderoiste @veprbl ready for review I think

Boop!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2817

SomeoneSerge commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -1,243 +1,54 @@
+          pythonPackages = packages."blender_${version}".pythonPackages;
+        };
+
+        inherit jackaudioSupport hipSupport cudaSupport colladaSupport spaceNavSupport waylandSupport;
```
You don't have to manually pass `cudaSupport`&c into `callPackage` arguments, you can just move all of the `config` and `isLinux` refernces to your `generic.nix`

SomeoneSerge commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+, openvdb, libXxf86vm, tbb, alembic
+, zlib, zstd, fftw, opensubdiv, freetype, jemalloc, ocl-icd, addOpenGLRunpath
+, jackaudioSupport, libjack2
+, cudaSupport, cudaPackages ? { }
```
`cudaPackages` is always defined, no need for `? { }`

the-furry-hubofeverything replied:
This is part of the original file. Will remove.

SomeoneSerge commented:
Sorry, I started reviewing and then realized the diff is too large for me to handle today

Madouura reviewed (commented):
Mainly checking splicing as per #211340.
If attr sets are not supposed to be spliced, but their attributes are, ~~then LGTM as far as that goes~~.
EDIT: ``blender-with-packages`` may need to be revised, as it is not spliced.
https://github.com/NixOS/nixpkgs/blob/c91c46ddf11311fe334da3354edca51b7eac9ac9/pkgs/applications/misc/blender/default.nix#L30-L33
```console
nix-repl> np.__splicedPackages.blender ? __spliced                                                   
true
nix-repl> np.__splicedPackages.blender-lts ? __spliced           
true
nix-repl> np.__splicedPackages.blender-hip ? __spliced           
true
nix-repl> np.__splicedPackages.blender-hip-lts ? __spliced       
true
nix-repl> np.__splicedPackages.blender-with-packages ? __spliced                                     
false
nix-repl> np.__splicedPackages.blenderVersions ? __spliced       
false
nix-repl> np.__splicedPackages.blenderVersions.stable ? __spliced
true
nix-repl> np.__splicedPackages.blenderVersions.lts ? __spliced    
true
```

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
````diff
@@ -440,6 +440,22 @@ The module update takes care of the new config syntax and the data itself (user
 
+- `blender` now provides both the lastest stable and LTS versions under `blenderVersions`. `blender` and `blender-hip` now default to `blenderVersions.stable`, which points to the latest stable version. The latest LTS version is available with `blender-lts`. Explicitly versioned packages are available with `blenderVersions.blender_3_3` and `blenderVersions.blender_3_6`, as well as their `blender- ...
+
+```
````
Please remove this site listing, it probably gets out of date over time and wrong documentation is worse than none.

the-furry-hubofeverything replied:
It's a release note? It might change yes, but for this specific version of NixOS it's true

You addressed minor versions in another comment, that's a good point, I removed the mentions

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -440,6 +440,22 @@ The module update takes care of the new config syntax and the data itself (user
 
 - `fusuma` now enables the following plugins: [appmatcher](https://github.com/iberianpig/fusuma-plugin-appmatcher), [keypress](https://github.com/iberianpig/fusuma-plugin-keypress), [sendkey](https://github.com/iberianpig/fusuma-plugin-sendkey), [tap](https://github.com/iberianpig/fusuma-plugin-tap) and [wmctrl](https://github.com/iberianpig/fusuma-plugin-wmctrl).
 
+- `blender` now provides both the lastest stable and LTS versions under `blenderVersions`. `blender` and `blender-hip` now default to `blenderVersions.stable`, which points to the latest stable version. The latest LTS version is available with `blender-lts`. Explicitly versioned packages are available with `blenderVersions.blender_3_3` and `blenderVersions.blender_3_6`, as well as their `blender- ...
```
We should not advertise explizit minor versions as they are going to be removed without aliases.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -481,3 +497,5 @@ The module update takes care of the new config syntax and the data itself (user
 
 - `teleport` has been upgraded from major version 12 to major version 14. Please see upstream [upgrade instructions](https://goteleport.com/docs/management/operations/upgrading/) and release notes for versions [13](https://goteleport.com/docs/changelog/#1300-050823) and [14](https://goteleport.com/docs/changelog/#1400-092023). Note that Teleport does not officially support upgrades across more th ...
+
+- All `blender` packages now passes its `pythonPackages` under `blender.pythonPackages`.
```
Please follow the comment under the next headline and randomly sort this.

SuperSandro2000 commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -1,243 +1,54 @@
-, openpgl
-, mesa
-, runCommand
+{ config, lib, stdenv, llvmPackages_10, darwin, callPackage, python310Packages
```
Please format one per line

SuperSandro2000 commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -1,243 +1,54 @@
-
+  versions = lib.importJSON ./versions.json;
+
+  latestStable = lib.last (builtins.sort lib.versionOlder (builtins.attrNames versions));
```
Please be explizit about this. The last version does not need to be stable, it could also be nightly/unstable/etc

the-furry-hubofeverything replied:
Well, the update script only fetches actively maintained stable versions, and feeds into `blenderVersion.stable`. Maintaining nightly/unstable versions is out of scope.

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+}:
+
+let
+  python = pythonPackages.python;
```
It is probably easier to just pass in python and use python.pkgs

the-furry-hubofeverything replied:
Part of original file, but will do

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+  optix = fetchzip {
+    # url taken from the archlinux blender PKGBUILD
+    url = "https://developer.download.nvidia.com/redist/optix/v7.3/OptiX-7.3.0-Include.zip";
+    sha256 = "0max1j4822mchj0xpz9lqzh91zkmvsn4py0r174cvqfz8z8ykjk8";
```
Please use Sri hashes

the-furry-hubofeverything replied:
This was part of the original file. I'll change it though

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+let
+  python = pythonPackages.python;
+  optix = fetchzip {
+    # url taken from the archlinux blender PKGBUILD
```
Link?

the-furry-hubofeverything replied:
This was part of the original file. I can still find it if it's necessary.

the-furry-hubofeverything replied:
Arch also has updated to Optix 7.6. I don't understand the differences between the versions, but Blender build instructions currently states to use 7.3, so I'm leaving it as 7.3.

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+    rev = "v${version}";
+    hash = hashes.addons;
+
+    postFetch = "patch -p3 -d $out < ${./draco-p2.patch}";
```
```suggestion
    postFetch = ''
      patch -p3 -d $out < ${./draco-p2.patch}
    '';
```

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+
+in
+stdenv.mkDerivation (finalAttrs: rec {
+  inherit version;
+
+  pname = "blender";
```
```suggestion
  pname = "blender";
  inherit version;
```

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+    rev = "v${version}";
+    hash = hashes.blender;
+
+    # Blender 3.3 uses submodules, but 3.5+ doesnt.
+    # See https://projects.blender.org/blender/blender/pulls/104755
+    fetchSubmodules = if lib.versionOlder version "3.5" then true else false;
```
```suggestion
    # Blender 3.3 uses submodules, but 3.5+ doesnt.
    # See https://projects.blender.org/blender/blender/pulls/104755
    fetchSubmodules = lib.versionOlder version "3.5";
```
having this here is kinda awkward but moving it into the json is also kinda awkward

the-furry-hubofeverything replied:
My opinion is that it's an awkward change overall. Blender expects people to use `make` and various scripts to fetch the dependencies, which doesn't really work with this.

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+    fetchSubmodules = if lib.versionOlder version "3.5" then true else false;
+  };
+
+  prePatch =
+    if lib.versionAtLeast version "3.5" then ''
+      mkdir scripts/addons
+      cp -Rv ${blender-addons}/* scripts/addons
+    '' else ''
+      cp -Rv ${blender-addons}/* release/scripts/addons
+    '';
```
```suggestion
  prePatch = if lib.versionAtLeast version "3.5" then ''
    mkdir scripts/addons
    cp -Rv ${blender-addons}/* scripts/addons
  '' else ''
    cp -Rv ${blender-addons}/* release/scripts/addons
  '';
```

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+    ./draco-p1.patch
+  ] ++ lib.optional stdenv.isDarwin ./darwin.patch;
+
+  nativeBuildInputs =
+    [ cmake makeWrapper pythonPackages.wrapPython llvmPackages.llvm.dev
+    ]
+    ++ lib.optionals cudaSupport [ addOpenGLRunpath ]
```
```suggestion
  nativeBuildInputs = [
    cmake
    makeWrapper
    pythonPackages.wrapPython
    llvmPackages.llvm.dev
  ] ++ lib.optionals cudaSupport [ addOpenGLRunpath ]
```
if we are refactoring we might as well use consistent one per line style

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+    ++ lib.optional spaceNavSupport libspnav;
+  pythonPath = with pythonPackages; [ numpy requests zstandard ];
+
+  postPatch = ''
+  '' +
+    (if stdenv.isDarwin then ''
```
```suggestion
  postPatch = (if stdenv.isDarwin then ''
```

SuperSandro2000 commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+      substituteInPlace extern/clew/src/clew.c --replace '"libOpenCL.so"' '"${ocl-icd}/lib/libOpenCL.so"'
+    '') +
+    (lib.optionalString hipSupport ''
+      substituteInPlace extern/hipew/src/hipew.c --replace '"/opt/rocm/hip/lib/libamdhip64.so"' '"${rocmPackages.clr}/lib/libamdhip64.so"'
+      substituteInPlace extern/hipew/src/hipew.c --replace '"opt/rocm/hip/bin"' '"${rocmPackages.clr}/bin"'
```
```suggestion
      substituteInPlace extern/hipew/src/hipew.c \
        --replace '"/opt/rocm/hip/lib/libamdhip64.so"' '"${rocmPackages.clr}/lib/libamdhip64.so"' \
        --replace '"opt/rocm/hip/bin"' '"${rocmPackages.clr}/bin"'
```

the-furry-hubofeverything commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+      "-DWITH_OPENSUBDIV=ON"
+      "-DPYTHON_LIBRARY=${python.libPrefix}"
+      "-DPYTHON_LIBPATH=${python}/lib"
+      "-DPYTHON_INCLUDE_DIR=${python}/include/${python.libPrefix}"
```
```
error: attribute 'includeDir' missing

       at /nix/store/szlv9cpcr7s5l013i9wq254r5q6qppby-source/pkgs/applications/misc/blender/derivation.nix:153:31:

          152|       "-DPYTHON_LIBPATH=${python}/lib"
          153|       "-DPYTHON_INCLUDE_DIR=${python.includeDir}"
             |                               ^
          154|       "-DPYTHON_VERSION=${python.pythonVersion}"
 ```

the-furry-hubofeverything commented on pkgs/applications/misc/blender/derivation.nix:
```diff
@@ -0,0 +1,280 @@
+  # libstdc++ in our RPATH. Sigh.
+  NIX_LDFLAGS = lib.optionalString cudaSupport "-rpath ${stdenv.cc.cc.lib}/lib";
+
+  blenderExecutable =
+    placeholder "out" + (if stdenv.isDarwin then "/Applications/Blender.app/Contents/MacOS/Blender" else "/bin/blender");
```
Results in build failure 

```
Builder called die: Cannot wrap '/1qgr252wjqadgkdd2sp04r4fqlrjzadd6mmcl50qjrd6mk3pil4pbin/blender' because it is not an executable file
```

SomeoneSerge replied:
You had it right, `placeholder` takes the name of the output, which is `"out". Although I think it's nice to avoid it when possible

the-furry-hubofeverything replied:
> You had it right, `placeholder` takes the name of the output, which is `"out". Although I think it's nice to avoid it when possible

@SomeoneSerge This is also part of the original file, I'm not exactly sure how to fix it.

the-furry-hubofeverything commented:
Squashed commits, addressed changes from @SomeoneSerge. 

@SuperSandro2000 I've implemented many changes requested, but I'm confused with some parts, and I left some comments for you.

@Madouura I have no idea with splicing is. It seems related to how `blender-with-packages` is implemented. With this PR,  `blender-with-packages` would be moved as an alias, and would be pointed at `blenderVersions.stable.withPackages` instead.

I've only done the bare minimum to make `wrapper.nix` (where withPackages is defined) support my package implementation. Any problems specifically with the implementation of `wrapper.nix` is out of scope for this PR, since I didn't make it, it isn't exactly part of my original goals, and I don't know how to fix it.

I guess feel free to make an issue or PR?

the-furry-hubofeverything commented:
@lucasew current changes might've broken `bpycv`, sorry. I'll have to deal with that after sleep.

the-furry-hubofeverything commented:
Right, everything should work now @lucasew

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

lucasew commented:
The test is working!!

![0](https://github.com/NixOS/nixpkgs/assets/15693688/043c97eb-f5c5-40b3-876f-f1e2d55c0c3d)

the-furry-hubofeverything commented:
# Potential breaking change

`wrapper.nix` is weird. It doesn't work exactly like `python.withPackages`, you can't do:
```
[hubble@Gulo-Laptop:/run/media/hubble/Data/GithubFurry/nixpkgs]$ nix-build -I nixpkgs=./ -E 'with (import <nixpkgs> {}); blender.withPackages(ps: [ps.bpycv])'
error: value is a function while a set was expected

       at «string»:1:29:

            1| with (import <nixpkgs> {}); blender.withPackages(ps: [ps.bpycv])
             |                             ^
```

`blender-with-packages` takes a list of python modules, but it doesn't pass the python packages. Taking a look from the python implmentation for `with-packages.nix`, I rewrote it a bit:
```
[hubble@Gulo-Laptop:/run/media/hubble/Data/GithubFurry/nixpkgs]$ nix-build -I nixpkgs=./ -E 'with (import <nixpkgs> {}); blender.withPackages(ps: [ps.bpycv])'
/nix/store/af8dzhz5b51l7dmwrn3klsldfvpzakwq-blender-wrapped-3.6.5
```

You can still rename the wrapped version -
```
[hubble@Gulo-Laptop:/run/media/hubble/Data/GithubFurry/nixpkgs]$ nix-build -I nixpkgs=./ -E 'with (import <nixpkgs> {}); blender.withPackages(ps: [ps.bpycv]){name = "test";}'
/nix/store/3r8c72cqzpdmfrfv9p8i2agy7ad5dl8n-blender-test-3.6.5
```

This will allow the use for proper versioned `pythonPackages` for the specific blender versions. but it will break backwards compatibility with `blender-with-packages`. **If this is the case, we should aim to merge the PR before the 30th.**

@lucasew @SuperSandro2000 Thoughts?

lucasew commented:
The idea of blender-with-packages is to always use the correct python pkgs scope according to the blender version. If the current implementation doesn't do this and the new does the new should be used, even if it breaks something.

I think it would be useful to add this detail in the release notes tho.

the-furry-hubofeverything commented:
https://github.com/the-furry-hubofeverything/nixpkgs/commit/bf01de25f74ad1a89a5b875ba0e26229478386d5

Updated to include `bpycv` render test fixup and documentation change

The fork is https://github.com/the-furry-hubofeverything/nixpkgs/tree/blender-withPackages-test

the-furry-hubofeverything commented:
you know what, I'm going to put it in

the-furry-hubofeverything commented:
Edited the main message, but I'll repeat it here.

**I moved a lot of the original code, but I kept it the same for the most part. I'm new to writing for nixpkgs, so I'm only implementing the parts I know and a major refactor of the original code is not within scope.**

SomeoneSerge commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -1,243 +1,51 @@
-      addOpenGLRunpath "$program"
-    done
-  '';
+        withPackages = f: let packages = f value.python.pkgs; in buildEnv.override { blender = value; extraModules = packages; };
```
I suspect that if one defines

```nix
anotherBlender = oldBlender.override { ... };
wrapped = anotherBlender.withPackages (ps: [ ... ]);
```

...then `wrapper.nix` might actually be called with `blender = oldBlender`, instead of `blender = anotherBlender`

the-furry-hubofeverything replied:
I'm confused, you mean that is wouldn't respect the override?

SomeoneSerge replied:
Yes, you pass the `value`, but `anotherBlender` is `value.override { ... }` (I think)

SomeoneSerge replied:
You could maybe just move this definition to `derivation.nix` and use `finalAttrs` instead of `blender`

SomeoneSerge replied:
But also maybe I was wrong, so check first

the-furry-hubofeverything replied:
@SomeoneSerge I don't think you can even override python

the-furry-hubofeverything replied:
you're right, the override isn't used. I'll try to tfix

SomeoneSerge commented on pkgs/applications/misc/blender/wrapper.nix:
```diff
@@ -2,17 +2,15 @@
-}:
-{ name ? "wrapped"
-, packages ? []
+, extraModules ? []
```
Nit: here in a few other places the spaces in brackets and parentheses don't seem to be used consistently? Like, sometimes it's `[ a ]`, and sometimes it's `[a]` within the same file. I think it's usually ok to mimic/preserve whatever style the file has been using

the-furry-hubofeverything replied:
Original file had no spaces in empty brackets, but do have spaces at each end for lists in the original file. Not changing.

SomeoneSerge reviewed (commented):
@the-furry-hubofeverything, this is some great work, thank you!

Personally, I'd like to go through the changes in more detail. I see that github is very confused about moving `default.nix` to `derivation.nix`, both in the [branch diff](https://github.com/NixOS/nixpkgs/compare/master...the-furry-hubofeverything:nixpkgs:blender-improvements) and the [commit](https://github.com/NixOS/nixpkgs/pull/257780/commits/14c39986cde386bffaf48a993d6b0b9bf03fa10a) views. I tried to insert a [commit](https://github.com/SomeoneSerge/nixpkgs/commit/41ac9dd465f6ebc7b2d84b3f850112487735c1d3) that moves `default.nix` without creating the new file. This [fixed the diff in the commit view](https://github.com/SomeoneSerge/nixpkgs/commit/f69fec7557169c0daa3e81a0ef188cf99d27cd1c), but not [the diff for the branch](https://github.com/SomeoneSerge/nixpkgs/compare/master...SomeoneSerge:nixpkgs:blender-improvements?expand=1). I can't say if this is worth it but maybe if you try this trick I could attach comments to specific commits (they'll still look broken in the merged view). This also seems to fix `git blame` (at least for me locally). Maybe there are better ways though

RE: `blender-with-packages = throw ...`

👍🏻 I think it's for the best, let the people update manually

It's also maybe worth explicitly noting that with this change Hydra will be building quite a few more versions of `blender` (which imo is a good thing)

Madouura commented:
Quick note: If you're going to ``throw`` ``blender-with-packages``, move it to ``top-level/aliases.nix``.

the-furry-hubofeverything commented on pkgs/applications/misc/blender/derivation.nix:
Considering what I saw with the commit view, I'm going to put in the `default.nix -> generic.nix` commit like you did

the-furry-hubofeverything commented:
> Quick note: If you're going to `throw` `blender-with-packages`, move it to `top-level/aliases.nix`.

That is already done with the latest commit.

the-furry-hubofeverything commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -1,243 +1,51 @@
+, llvmPackages_10
+, darwin
+, callPackage
+, python310
```
> When the user wants to override python, should they `(blenderVersions.override { python = ...; }.stable` or `blender.override { python = ...; }`?

@SomeoneSerge Since `blender` is just an alias for `blenderVersions.stable`, `blender.override { python = ...; }` is the same as `blenderVersions.stable.override { python = ...; }`

the-furry-hubofeverything replied:
Actually, I'm not even sure you can override python

the-furry-hubofeverything replied:
@SomeoneSerge I don't think you can even override python for this one

SomeoneSerge replied:
Well, my thinking was: ok, both `default.nix` and `generic.nix` take `python` as an argument (now `python310Packages` and `pythonPackages` resp.), so I can override either of them if I want to patch python. I probably want to do the most local override I can, i.e. `blender.override`. But is it safe, and how do I know whether `blender` consumes something else from `blenderVersions` that might have remembered the old python? (the answer is to read the code obviously)

Although now both should work, so maybe anything goes

SomeoneSerge replied:
(moving the conversation back from [matrix](https://matrix.to/#/%23dev%3Anixos.org/%24N4UWVWsm6lh72clnVbJ-ongpUtbs77n8owzhty1WkM8?via=someonex.net&via=matrix.org&via=snug.si&via=turbo.ooo)) What you could do is remove any mentions of python from `default.nix`, and take both `python310Packages` and `pythonPackages ? python310Packages` in `generic.nix`

But I'm not saying you should. This is susceptible to yak shaving

the-furry-hubofeverything replied:
> Well, my thinking was: ok, both `default.nix` and `generic.nix` take `python` as an argument (now `python310Packages` and `pythonPackages` resp.), so I can override either of them if I want to patch python. I probably want to do the most local override I can, i.e. `blender.override`. But is it safe, and how do I know whether `blender` consumes something else from `blenderVersions` that might have remembered the old python? (the answer is to read the code obviously)
> 
> Although now both should work, so maybe anything goes

(didn't notice sorry)

All the user needs to worry about is to override the package itself, not `blenderVersions`, in the same vein as `minecraftServers`, where it uses `javaPackages` as inputs grabs specific versions of JRE for its `derivation.nix`

I don't think there's anything I could do.

the-furry-hubofeverything replied:
@SomeoneSerge ie unless `generic.nix` should have dependency logic (which version of dependency for which), I don't think I can avoid this

the-furry-hubofeverything committed:
blender: default.nix -> generic.nix

Git/Github seems to be confused for this rename, so a separate commit is needed.

the-furry-hubofeverything commented:
All right!

- renamed "derivation.nix" to "generic.nix", and used dedicated commit to rename (since Github couldn't pick it up)
- move from using `python` as argument to `pythonPackages`, and removed all instances of `python.pkgs` 
  - as `python` is defined as `pythonPackages.python`, the python version is dependent on `pythonPackages`, so conflicts should not happen
- moved `withPackages` from default.nix to generic.nix (using finalAttrs) to respect overrides
- while I was at it, I removed "rec" in favour of "finalAttrs". 

@SomeoneSerge @SuperSandro2000

the-furry-hubofeverything commented:
Following a comment from Figsoda, I'll be moving the breaking changes to a separate PR. This will be converted back to draft until that PR is merged (hopefully before the release freeze, if not I'll put it back in)

lucasew commented:
Result of `nixpkgs-review pr 257780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender (blender-lts ,blenderVersions.blender_3_6 ,blenderVersions.lts ,blenderVersions.stable)</li>
    <li>blender-hip (blender-hip-lts ,blenderVersions.blender-hip_3_6)</li>
    <li>blenderVersions.blender-hip_3_3</li>
    <li>blenderVersions.blender_3_3</li>
  </ul>
</details>

the-furry-hubofeverything commented:
Just let me know if I need to do anything

lizelive commented:
one important thing that is missing from nix's blender is https://docs.blender.org/api/current/info_advanced_blender_as_bpy.html

the-furry-hubofeverything commented:
Why did github actions add so many actions what the

the-furry-hubofeverything commented:
Would it be easier to review this PR if it were to be split up to several smaller ones?

SharzyL commented:
It seems that you need to rebase your commits

the-furry-hubofeverything commented:
Dear lord-

the-furry-hubofeverything commented:
I'll split it up when I get the time to do so. Will set to draft (and maybe eventually close) since 

1. The two commits that need to be squashed ("blender: implement multi version packages")
2. Blender 3.3 doesn't build on opencolorio 2.3 :anguished: 
3. It's getting very big and messy

SuperSandro2000 commented on pkgs/applications/misc/blender/generic.nix:
```diff
@@ -0,0 +1,293 @@
+  '';
+
+  passthru = {
+    inherit python pythonPackages isLTS;
```
```suggestion
    inherit python isLTS;
```
we should only inherit python from where you can get pkgs via python.pkgs

the-furry-hubofeverything replied:
Re: https://github.com/NixOS/nixpkgs/pull/257780#pullrequestreview-1698491552

This was implemented over `python3.pkgs` as per #211340

SuperSandro2000 commented on pkgs/applications/misc/blender/generic.nix:
```diff
@@ -0,0 +1,293 @@
+          echo "Rendering with $engine..."
+
+          # Blender doesn't support headless rendering on EEVEE before 3.4
+          if [ ${if lib.versionOlder version "3.4" then "0" else "1"} -eq 0 ] && [ $engine == "BLENDER_EEVEE" ]
```
```suggestion
          if ${lib.boolToString (lib.versionOlder version "3.4")} && [ $engine == "BLENDER_EEVEE" ]
```

SuperSandro2000 reviewed (commented):
mostly formatting and design things

the-furry-hubofeverything commented:
Closing, as this got really messy and it could use a bit of a clean up/rewrite

continuing with smaller, easier to review PRs

the-furry-hubofeverything commented:
Work continues on #305727
